File: day02/part2.py
from common import main
from day02.input import parse_input

MIN_REPEAT = 2

def explore_patterns(start: int, end: int, pattern_length: int) -> set[int]:
    start_str = str(start)
    start_length = len(start_str)
    end_length = len(str(end))

    pattern_repetition_counts = [
        n // pattern_length
        for n in range(start_length, end_length + 1)
        if n % pattern_length == 0 and n > pattern_length
    ]
    if len(pattern_repetition_counts) == 0:
        return set()
    
    invalid_ids = set()
    
    for pattern_repetitions in pattern_repetition_counts:
        if pattern_repetitions * pattern_length == start_length:
            # Start with starting pattern in `start`.
            candidate_pattern = start_str[:pattern_length]
            candidate_int = int(candidate_pattern)
        else:
            # Start at the beginning for this number of sigfigs.
            candidate_pattern = "1" + ("0" * (pattern_length - 1))
            candidate_int = int(candidate_pattern)

        candidate = int(candidate_pattern * pattern_repetitions)
        while candidate <= end:
            if candidate >= start:
                invalid_ids.add(candidate)

            # Move to next pattern, checking for wraparound.
            candidate_int += 1
            candidate_pattern = str(candidate_int)
            if len(candidate_pattern) > pattern_length:
                # Overflow!
                break

            candidate = int(candidate_pattern * pattern_repetitions)

    return invalid_ids
def find_invalid_ids(start: int, end: int) -> set[int]:
    end_str = str(end)

    # Set comprehension to avoid duplicates.
    return {
        invalid_id
        for pattern_length in range(1, (len(end_str) // MIN_REPEAT) + 1)
        for invalid_id in explore_patterns(start, end, pattern_length)
    }

def solve_part2(id_ranges: list[tuple[int, int]]):
    return sum([
        sum(find_invalid_ids(start, end))
        for start, end in id_ranges
    ])

if __name__ == "__main__":
    main(parse_input, solve_part2)

File: day02/part2.rs
use std::collections::HashSet;

mod input;

use common::run;
use input::{parse_input, Input};

fn explore_patterns(start: usize, end: usize, pattern_length: usize) -> HashSet<usize> {
    let start_str = start.to_string();
    let start_length = start_str.len();
    let end_length = end.to_string().len();

    (start_length..=end_length)
        .filter(|&n| n % pattern_length == 0 && n > pattern_length)
        .map(|n| n / pattern_length)
        .flat_map(|pattern_repetitions| {
            let mut invalid_ids = vec![];

            let (mut candidate_pattern, mut candidate_int) = if pattern_repetitions * pattern_length == start_length {
                let pat = &start_str[..pattern_length];
                (pat.to_string(), pat.parse::<usize>().unwrap())
            } else {
                let pat = format!("1{}", "0".repeat(pattern_length - 1));
                (pat.clone(), pat.parse::<usize>().unwrap())
            };

            let mut candidate = candidate_pattern.repeat(pattern_repetitions).parse::<usize>().unwrap();

            while candidate <= end {
                if candidate >= start {
                    invalid_ids.push(candidate);
                }

                candidate_int += 1;
                candidate_pattern = candidate_int.to_string();
                if candidate_pattern.len() > pattern_length {
                    break;
                }
                candidate = candidate_pattern.repeat(pattern_repetitions).parse::<usize>().unwrap();
            }
            invalid_ids
        })
        .collect()
}

fn find_invalid_ids(start: usize, end: usize) -> HashSet<usize> {
    let end_str = end.to_string();
    let longest_pattern = end_str.len() / 2;

    (1..=longest_pattern).into_iter()
        .flat_map(|pattern_length| explore_patterns(start, end, pattern_length))
        .collect()
}

fn solve_part2(id_ranges: Input) -> i64 {
	id_ranges.into_iter()
		.flat_map(|(start, end)| find_invalid_ids(start, end))
		.sum::<usize>() as i64
}

fn main() {
	run(parse_input, solve_part2);
}
